k8s-deploy/pkg/job/job.go

```diff
@@ -202,30 +186,49 @@ func ClusterUpdate(clusterArgs *modules.ClusterArgs, creator string) (*modules.J
 	if cluster.ChartName != clusterNew.ChartName {
 		return nil, fmt.Errorf("doesn't support upgrade between different charts")
 	}
-	if clusterNew.ChartName == fateChartName {
-		err = validateFateVersion(c.ChartVersion, specNew["imageTag"].(string))
-		if err != nil {
-			return nil, fmt.Errorf("the image tag is not consistent with the chart verison, which is unsupported")
-		}
-	}
 
 	if reflect.DeepEqual(specOld, specNew) &&
 		cluster.ChartName == clusterArgs.ChartName &&
 		cluster.ChartVersion == clusterArgs.ChartVersion {
 		return nil, fmt.Errorf("the configuration file did not change")
 	}
-	var upgradeScripts []string
-	if clusterNew.ChartName == fateChartName {
-		if cluster.ChartVersion != clusterArgs.ChartVersion {
-			upgradeScripts, err = getUpgradeScripts(cluster.ChartVersion, clusterArgs.ChartVersion)
-			if err != nil {
-				return nil, err
+
+	oldVersion, err := version.NewVersion(strings.ReplaceAll(cluster.ChartVersion, "v", ""))
+	newVersion, err := version.NewVersion(strings.ReplaceAll(clusterArgs.ChartVersion, "v", ""))
+	if newVersion.LessThan(oldVersion) {
+		return nil, fmt.Errorf("using kubefate to downgrade cluster is not supported yet")
+	}
+	// FmlFrameworkNameToUmStuffMap is a map, whose keys are the fml framework names, such as fate.
+	// Its values are the Information of the corresponding upgrade manager's chart.
+	// In the future we may extend this map to add more other fml frameworks
+	FmlFrameworkNameToUmInfoMap := modules.MapStringInterface{
+		fateChartName: modules.MapStringInterface{
+			"chartName":             fateUpgradeManagerChartName,
+			"chartVersion":          "v1.0.0",
+			"specConstructFunction": ConstructFumSpec,
+			"validationFunction":    validateFateVersion,
+		},
+	}
+	var upgradeManagerChartName string
+	if newVersion.GreaterThan(oldVersion) {
+		umInfo, ok := FmlFrameworkNameToUmInfoMap[clusterNew.ChartName]
+		if !ok {
+			log.Warn().Msgf("the fml framework %s doesn't have an upgrade manager", clusterNew.ChartName)
+		} else {
+			umSpecFuncName := reflect.ValueOf(umInfo.(modules.MapStringInterface)["validationFunction"])
+			params := []reflect.Value{reflect.ValueOf(cluster.ChartVersion),
+				reflect.ValueOf(clusterArgs.ChartVersion),
+				reflect.ValueOf(specNew["imageTag"].(string))}
+			res := umSpecFuncName.Call(params)
+			errInterface := res[0].Interface()
+			if errInterface != nil {
+				return nil, errInterface.(error)
```

Comment:
This works. Just a thought is that we may be able to archive this using interface, instead of scattered functions and reflection.

I am thinking something like:

```
type upgradeManager interface {
	validate(startVersion, chartVersion, imageVersion string) error
	chartInfo() (chartName, chartVersion string)
	constructSpec(oldSpec, newSpec modules.MapStringInterface) modules.MapStringInterface
	// or directly
	// getCluster()  (modules.Cluster, error)
	isFinished() (bool, error)
}
```

And then you can implement FUM struct complying with this interface. And program our `ClusterUpdate` with this interface without leaking any "FATE" specific terms (like `fateUpgradeJobName` in line 283) into this function, and without hacking around using golang reflection.

But that is just a thought for maybe future improvements. I'm fine if current code works.

Reply:
Let me do this in the next pr.